buildsystem: centralise package sourcing

MilhouseVH opened the pull request:
This is a clean up of the build system so that all package sourcing takes place using a single method, `config/options <pkg-name>`, removing the need to source naked packages for a second time in `scripts/unpack`, `scripts/install` and `scripts/build`.

Helper functions have been added to unset all package functions, and to source the package itself (with appropriate error handling).

The `get_pkg_variable()` function has been updated to call `source_package()` directly rather than via `config/options`, as the former is being called from within the context of the latter so it should not be necessary to incur the overhead of sourcing `config/options` each time a package variable is requested.

Where necessary, packages referencing `PKG_*` variables that are initialised _after_ the package is sourced (eg. `PKG_BUILD`) have been updated to implement `configure_package()` where such variables will be available and valid.

`scripts/uninstall` is removed as it serves no current purpose - can be reinstated if required.

`scripts/build`, `scripts/install` and `scripts/unpack` have been reorganised to bail out early when called incorrectly, and remove redundant variables/code.

`scripts/build` will now execute `setup_toolchain_$STARGET` _after_ the package has been sourced, and any packages that override or reference variables configured by `setup_toolchain_$TARGET` (eg. `CC`, `CXX`, `CFLAGS`, `LDFLAGS`, `HOST_CC`, `HOST_CFLAGS` etc.) must now override/reference these variables in a suitable `pre`-function, ie. `pre_build_*`, `pre_configure_*`, `pre_make_*` etc.

`scripts/install` will now output the `CLR_INSTALL` build message _after_ dependencies have been installed, and not before.

antonlacon commented on scripts/unpack:
```diff
@@ -4,14 +4,14 @@
-. config/options $1
-
 if [ -z "$1" ]; then
   die "usage: $0 package_name"
```
die will be unknown until config/functions gets sourced, so

```
echo -e .... >&2
exit 1
```

Same issue `scripts/install` and `scripts/build`.

MilhouseVH replied:
ah yes... very true! will fix.

MilhouseVH commented:
I've made a few additions after further testing - to simplify reviewing I've added 6 new commits which I'll squash eventually.

MilhouseVH commented:
The `release` change is required because the `release` script is being sourced [here](https://github.com/LibreELEC/LibreELEC.tv/blob/5a21612/scripts/image#L284), and in the case of RPi/Amlogic (Rockchip is OK) it then sources `config/options $1` where `$1` is whatever value has been used to execute `scripts/image`... this fails because we are validating the package name now, so when running `make release" you'd have the error:
```
        root (0)
Number of gids 2
        root (0)
        unknown (81)
Running /home/neil/projects/LibreELEC.tv/packages/tools/bcm2835-bootloader/release
FAILURE: release/package.mk does not exist
Makefile:9: recipe for target 'release' failed
make: *** [release] Error 1
```

antonlacon commented on projects/Amlogic/bootloader/release:
```diff
@@ -3,7 +3,5 @@
-
 mkdir -p $RELEASE_DIR/3rdparty/bootloader
-cp -a $INSTALL/usr/share/bootloader/* $RELEASE_DIR/3rdparty/bootloader
+  cp -a $INSTALL/usr/share/bootloader/* $RELEASE_DIR/3rdparty/bootloader
```
Stray indent?

MilhouseVH replied:
Not in this case, it's consistent with the existing `release` files where we indent-within-mkdir

https://github.com/LibreELEC/LibreELEC.tv/blob/30ced21a40bd47ad86261812418c577277161b59/projects/Rockchip/bootloader/release

https://github.com/LibreELEC/LibreELEC.tv/blob/30ced21a40bd47ad86261812418c577277161b59/packages/tools/bcm2835-bootloader/release

See coding standards RFC: https://github.com/LibreELEC/LibreELEC.tv/pull/3055/files#diff-27d64c693629212f25ac5e5c03c2e2a2R29

MilhouseVH commented:
Squashed, and cleaned up. I've now completed a dozen clean builds (RPi/RPi2/Generic) without issue, and also built tvheadend42. Hopefully there won't be further changes... I'll kick off some Amlogic/Rockhip/WeTek builds overnight.

antonlacon commented on config/functions:
```diff
@@ -803,6 +802,108 @@ pkg_call_optional() {
+    cd "${opwd}"
+
+    [ -z "$PKG_SHORTDESC" ] && PKG_SHORTDESC="$PKG_NAME (autogenerated)"
+    [ -z "$PKG_LONGDESC" ] &&  PKG_LONGDESC="$PKG_NAME (autogenerated)"
```
```
PKG_SHORTDESC="${PKG_SHORTDESC:-${PKG_NAME} (autogenerated)}"
PKG_LONGDESC="${PKG_LONGDESC:-${PKG_NAME} (autogenerated)}"
```

antonlacon commented:
`Config/path` needs a copyright header. Looks good to me after that and the `config/functions` quibble.

MilhouseVH commented:
Thanks @antonlacon - copyright and quibble addressed.

MilhouseVH commented:
Added fixes for a few more packages that reference variables initialised by `setup_toolchain $TARGET`.
